okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.java

```diff
@@ -22,92 +22,207 @@
-
-  private final WebSocketWriter writer;
   private final WebSocketReader reader;
+  private final WebSocketListener readerListener;
+  /** True after a close frame was read by the reader. No frames will follow it. */
+  private boolean readerSawClose;
+
+  final WebSocketWriter writer;
+  /** True after calling {@link WebSocketWriter#writeClose(int, String)} to send a close frame. */
+  final AtomicBoolean writerClosed = new AtomicBoolean();
+
+  /** Guarded by itself. Must check {@link #isShutdown} before enqueuing work. */
+  private final Executor replier;
 
   /** True after calling {@link #close(int, String)}. No writes are allowed afterward. */
-  private volatile boolean writerSentClose;
+  private boolean senderSentClose;
   /** True after {@link IOException}. {@link #close(int, String)} becomes only valid call. */
-  private boolean writerWantsClose;
-  /** True after a close frame was read by the reader. No frames will follow it. */
-  private boolean readerSentClose;
+  private boolean senderWantsClose;
+
+  private final Response response;
+  private final String name;
 
-  /** True after calling {@link #close()} to free connection resources. */
-  private final AtomicBoolean connectionClosed = new AtomicBoolean();
+  /** Guarded by {@link #replier}. True after calling {@link #shutdown()}. */
+  private boolean isShutdown;
 
   protected RealWebSocket(boolean isClient, BufferedSource source, BufferedSink sink, Random random,
-      Executor replyExecutor, WebSocketListener listener, String url) {
-    this.replyExecutor = replyExecutor;
-    this.listener = listener;
-    this.url = url;
+      Executor replier, WebSocketListener readerListener, Response response, String name) {
+    this.readerListener = readerListener;
+    this.replier = replier;
+    this.response = response;
+    this.name = name;
 
-    writer = new WebSocketWriter(isClient, sink, random);
     reader = new WebSocketReader(isClient, source, this);
+    writer = new WebSocketWriter(isClient, sink, random);
   }
 
-  @Override public final void onMessage(ResponseBody message) throws IOException {
-    listener.onMessage(message);
+  ////// READER THREAD
+
+  /** Read and process all socket messages delivering callbacks to the supplied listener. */
+  public final void loopReader() {
+    try {
+      readerListener.onOpen(this, response);
+    } catch (Throwable t) {
+      Util.throwIfFatal(t);
+      replyToReaderError(t);
+      readerListener.onFailure(t, null);
+      return;
+    }
+
+    while (readMessage()) {
+    }
   }
 
-  @Override public final void onPing(final Buffer buffer) {
-    replyExecutor.execute(new NamedRunnable("OkHttp %s WebSocket Pong Reply", url) {
-      @Override protected void execute() {
-        peerPing(buffer);
-      }
-    });
+  /**
+   * Read a single message from the web socket and deliver it to the listener. Returns false when
+   * no more messages can be read.
+   */
+  final boolean readMessage() {
```

Comment:
Actually it reads a single control frame _or_ a message that may be one to many frames with control frames potentially interleaved... so `readFrame()` isn't terrible it's just not always strictly correct. Maybe `readFrames()` + some documentation?